README.md
# YouPod

Pour commencer à programmer YouPod, nous sommes partis d'un constat : 
Il est difficile pour un néophyte d'exporter simplement son podcast en vidéo. 

S'il est possible d'utiliser les logiciels de montages classiques, il faut reconnaître que le réaliser "à la main", c'est long... 
Il existe des utilitaires en ligne, ou des hébergeurs de podcasts qui le proposent directement dans leurs offres (comme [Ausha](https://www.ausha.co/), ou [Acast Open](https://open.acast.com/)) mais tous sont payants.

En réalité, ce n'est pas très compliqué : Il suffit de prendre l'audio, et une image non? 

On a quand même ajouté une petite boucle vidéo derrière pour que ça ressemble un peu plus à quelque chose.

[![Demo](https://img.youtube.com/vi/Lpa5UtjI9NE/0.jpg)](https://www.youtube.com/watch?v=Lpa5UtjI9NE)
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FBigaston%2Fyoupod.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2FBigaston%2Fyoupod?ref=badge_shield)

Source : https://robotsettondeuses.lepodcast.fr/



[![Demo](https://img.youtube.com/vi/H5IUxtKbapI/0.jpg)](https://www.youtube.com/watch?v=H5IUxtKbapI)

Source : https://aspacemr.lepodcast.fr/

L'idée de réaliser cet outil vient de [ce tuto d'Alliés Numériques](https://alliesnumeriques.org/tutoriels/publier-votre-podcast-sur-youtube-la-methode-classe-avec-ffmpeg/).

## Installation

**/!\ Si vous  voulez uniquement générer des vidéos, sans vous prendre la tête avec les installations, Podshows m'a permis d'héberger une instance publique de Youpod sur [app.youpod.io](https://app.youpod.io)**

Pour avoir votre propre instance de Youpod c'est simple : Il vous faudra simplement [NodeJS](https://nodejs.org/en/) (en version LTS). 
Une fois NodeJS installé, téléchargez le code, et installez les modules. 

Il est nécéssaire aussi d'avoir FFMPEG 4.2.x minimum. Verifiez en tapant dans un terminal
```ffmpeg -version```
Si ce n'est pas le cas, téléchargez-le ici : https://www.ffmpeg.org/download.html

Voici les commandes pour faire ça :
```shell
git clone https://github.com/Bigaston/youpod
cd youpod
npm install --unsafe-perm=true --allow-root
```

(Sur Windows la commande cd ne marchera pas, rendez vous simplement dans le dossier téléchargé, cliquez sur la barre d'adresse, écrivez `cmd` et vous aurez un terminal dans le bon dossier pour lancer `npm install`.

Ensuite il est très important de se rendre dans le fichier `/youpod/.env` ainsi généré, qui contiendra les informations essentielles à l'installation de votre site web.

```.env
PORT=5674                       # Le port de votre serveur
HOST=http://localhost:5674      # L'adresse à laquelle on pourra accèder à votre site
EXPORT_FOLDER=./video           # Le dossier où seront sauvegardés les vidéos
ADMIN_PWD=123456                # Le mot de passe du dashboard d'administration
GOOGLE_ID=liqusdhnqoi           # Votre client ID de l'API Youtube
GOOGLE_SECRET=qosidhqnpsodih    # Le code secret de l'API Youtube
```

Maintenant il ne vous reste plus qu'à lancer votre serveur avec

```shell
npm start
```

Et vous pourrez y accéder directement via un navigateur à l'adresse que vous avez configurée (par défaut, http://localhost:5674).

Maintenant il faut vous rendre sur votre site web, sur `/admin` pour terminer la configuration. Entrez votre mot de passe administrateur et descendez sur le bas de la page. Pensez à compléter tous les champs de vos options (sauf ceux facultatifs).

Pour plus d'information à propos de comment configurer votre compte email, allez voir dans [cet article du Wiki](https://github.com/Bigaston/youpod/wiki/Configurer-son-compte-mail) !

## Remerciement

Merci beaucoup à [Phil_Goud](https://twitter.com/Phil_Goud) qui m'a beaucoup aidé sur ce projet grâce entre autre aux vagues qu'il a généré et à toute l'interface qu'il a faite (vu que la mienne était un peu... bof?)

Merci à [Pof](https://twitter.com/PofMagicfingers) pour avoir boosté aux hormones notre commandes FFMPEG.

Si vous souhaitez me soutenir financièrement dans ce projet, vous pouvez passer par mon [Patreon](https://patreon.com/Bigaston) ou mon [uTip](https://utip.io/Bigaston) !

En cas de problème ou de questions, passez directement par les [Issues](https://github.com/Bigaston/youpod/issues) ou par [mon Twitter](https://twitter.com/Bigaston) !



## License
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FBigaston%2Fyoupod.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2FBigaston%2Fyoupod?ref=badge_large)

docs/changelog.md
# Changelog
## Version 1.2
### Version 1.2.4
#### Ajouts
- Modification du footer des pages web (ajout lien Twitter, Github et confidentialité)
- Vérification du MP3 pour les créations custom
- Vérification du MP3 pour les créations à partir de flux
- Changement de l'encodage (192kbps)

#### Corrections
- Correction bug d'affichage de l'image

### Version 1.2.3
#### Corrections
- Bug dans la recherche dans le flux quand l'épisode ne contenait pas d'enclosure
- Corrections fautes dans le mail
- Correction temps démarage extrait
- Suppression slider à cause de problèmes

### Version 1.2.2
#### Ajouts
- Clé de l'API google sur l'admin
- Lecteur pour définir le début de l'extrait à l'oreille
- Social Custom
- Option de police déplacée dans option avancée
- Social sur le dashboard admin
- Un des slider a été masqué

#### Corrections
- Changement sur le mail pour les social
- Bug lors de la lecture de l'extrait

### Version 1.2.1
#### Ajouts
- Preview pour les réseaux sociaux à partir du flux RSS

### Version 1.2.0
#### Ajouts
- Choix du service d'email entre SMTP et Gmail

## Version 1.1.0
#### Ajouts
- Changement au niveau de la gestion de la base de donnée vers Sequelize
- Ajout d'un panel d'administration
  - Configuration des options du site
  - Liste des vidéos en attentes et en traitement
  - Statistiques
- Page d'explication en cas de mauvais CSRF
- Sécurité anti spam (CSRF et vérification de certains paramètres de duplication)
- Footer définit grâce à un partial
- Ajout de l'upload directement sur Youtube
  - Activable/désactivable depuis l'admin
  - Configuration des clés d'accès dans le .env
  - Génération automatique du titre et de la description
  - Connection via le compte Google
  - Ajout d'une politique de confidentialité

#### Correction
- Définition du transporteur NodeMailer juste avant l'envoit du mail
- Clé d'API de Google Font Personalisable

install.js
const fs = require("fs")
const path = require("path")
const db = require(__dirname + "/models/index.js")

console.log("Création des fichiers de base")

if (!fs.existsSync(path.join(__dirname, "/.env"))) {
    fs.writeFileSync(path.join(__dirname, ".env"), `
PORT=5674
HOST=http://localhost:5674
EXPORT_FOLDER=./video
ADMIN_PWD=123456
COOKIE_SECRET=IDK`)
}

db.sequelize.sync().then(()=> {
    console.log("Création des tables réussies!")
    db.Option.findOrCreate({where: {key: 'MAX_DURING'}, defaults: {value: '1'}}).then(() => {
        db.Option.findOrCreate({where: {key: 'MAX_DURING_SOCIAL'}, defaults: {value: '1'}}).then(() => {
            db.Option.findOrCreate({where: {key: 'KEEPING_TIME'}, defaults: {value: '12'}}).then(() => {
                db.Option.findOrCreate({where: {key: 'MAIL_SERVICE'}, defaults: {value: 'gmail'}}).then(() => {
                    db.Option.findOrCreate({where: {key: 'SMTP_HOST'}, defaults: {value: 'send.example.com'}}).then(() => {
                        db.Option.findOrCreate({where: {key: 'SMTP_DOMAIN'}, defaults: {value: 'send.example.com'}}).then(() => { 
                            db.Option.findOrCreate({where: {key: 'SMTP_PORT'}, defaults: {value: '587'}}).then(() => { 
                                db.Option.findOrCreate({where: {key: 'SMTP_USERNAME'}, defaults: {value: 'default'}}).then(() => { 
                                    db.Option.findOrCreate({where: {key: 'SMTP_PASSWORD'}, defaults: {value: '123456'}}).then(() => { 
                                        db.Option.findOrCreate({where: {key: 'GMAIL_ADDR'}, defaults: {value: 'michel@example.com'}}).then(() => {
                                            db.Option.findOrCreate({where: {key: 'GMAIL_PWD'}, defaults: {value: '123456'}}).then(() => {
                                                db.Option.findOrCreate({where: {key: 'GEN_PWD'}, defaults: {value: ''}}).then(() => {
                                                    db.Option.findOrCreate({where: {key: 'API_PWD'}, defaults: {value: '123456'}}).then(() => {
                                                        db.Option.findOrCreate({where: {key: 'GOOGLE_FONT_KEY'}, defaults: {value: '123456'}}).then(() => {
															db.Option.findOrCreate({where: {key: 'GOOGLE_ID'}, defaults: {value: '123456'}}).then(() => {
																db.Option.findOrCreate({where: {key: 'GOOGLE_SECRET'}, defaults: {value: '123456'}}).then(() => {
																	db.Option.findOrCreate({where: {key: 'ENABLE_YOUTUBE'}, defaults: {value: 'false'}})
																})
															})
                                                        })
                                                    })
                                                })
                                            })
                                        })
                                    })
                                })
                            })
                        })  
                    })
                })
            })
        })
    })
}).catch(error => {
    console.log("Erreur lors de la création des tables!")
})

if(!fs.existsSync(path.join(__dirname, "/video"))) {
    fs.mkdirSync(path.join(__dirname, "/video"))
}

if (!fs.existsSync(path.join(__dirname, "/tmp"))) {
    fs.mkdirSync(path.join(__dirname, "/tmp"))
}
console.log("L'instalation est terminée.")
console.log("Pensez aussi à éditer vos informations dans .env")